enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/general_permissions.clj

```diff
@@ -0,0 +1,126 @@
+  (if (perms/set-has-full-permissions? permissions-set (general-perms-path perm-type))
+    :yes
+    :no))
+
+(s/defn permissions-set->general-perms :- GroupPermissionsGraph
+  "Get a map of all general permissions for a group"
+  [permission-set]
+  {:setting      (perrmission-for-type permission-set :setting)
+   :monitoring   (perrmission-for-type permission-set :monitoring)
+   :subscription (perrmission-for-type permission-set :subscription)})
+
+(s/defn graph :- GeneralPermissionsGraph
+  "Fetch a graph representing the general permissions status for every group. This works just like the function of
+  the same name in `metabase.models.permissions`; see also the documentation for that function."
+  []
+  {:revision (g-perm-revision/latest-id)
+   :groups   (into {} (for [[group-id perms] (group-id->permissions-set)]
+                        {group-id (permissions-set->general-perms perms)}))})
+
+;;; -------------------------------------------------- Update Graph --------------------------------------------------
+
+(defn- save-perms-revision!
+  "Save changes made to the general permissions graph for logging/auditing purposes.
+  This doesn't do anything if `*current-user-id*` is unset (e.g. for testing or REPL usage)."
+  [current-revision old changes]
+  (when *current-user-id*
+    (db/insert! GeneralPermissionsRevision
+                ;; manually specify ID here so if one was somehow inserted in the meantime in the fraction of a second since we
+                ;; called `check-revision-numbers` the PK constraint will fail and the transaction will abort
+                :id      (inc current-revision)
+                :before  old
+                :changes changes
+                :user_id *current-user-id*)))
+
+(defn update-general-permissions!
+  "Perform update general permissions for a group-id"
+  [group-id changes]
+  (doseq [[permission-type permission-value] changes]
+    (case permission-value
+      :yes
+      (perms/grant-permissions! group-id (general-perms-path permission-type))
+
+      :no
+      (perms/delete-related-permissions! group-id (general-perms-path permission-type)))))
+
+(s/defn update-graph!
+  "Update the General Permissions graph.
+  This works just like [[metabase.models.permission/update-data-perms-graph!]], but for General permissions;
+  refer to that function's extensive documentation to get a sense for how this works."
+  [new-graph :- GeneralPermissionsGraph]
+  (let [old-graph          (graph)
+        old-perms          (:groups old-graph)
+        new-perms          (:groups new-graph)
+        ;; filter out any groups not in the old graph
+        new-perms          (select-keys new-perms (keys old-perms))
+        ;; filter out any collections not in the old graph
+        new-perms          (into {} (for [[group-id collection-id->perms] new-perms]
+                                      [group-id (select-keys collection-id->perms (keys (get old-perms group-id)))]))
```

Comment:
ah, I didn't change the name. but I think the gist is still valid for general permisison, because it's to prevent:
- accidiently grant a new permission that we don't know about
- grant for permisisons a new group

Reply:
wouldn't a group with no general permissions not show up in the old graph? but we still want to be able to set permissions for it

and collections shouldn't be relevant here at all, right? so I don't get what the last line is doing

Reply:
ah, it actually relates to our other comment about how to get group-ids->permisison-set.
Currently if a group has no general permissions at all, we still have in the graph